csrc/torch_memory_saver.cpp

```
@@ -211,78 +212,114 @@ class TorchMemorySaver {
+        std::cout << "[torch_memory_saver.cpp] tms_region_leave instance=" << this << std::endl;
 #endif
         is_interesting_region_ = false;
     }
 
-    bool is_interesting_region() {
+    bool is_interesting_region() const {
         return is_interesting_region_;
     }
+
+private:
+    std::mutex allocator_metadata_mutex_;
+    std::unordered_map<void *, _AllocationMetadata> allocation_metadata_;
+    bool is_interesting_region_;
+};
+
+// ----------------------------------------------- region manager ------------------------------------------------
+
+namespace RegionManager {
+    static thread_local TorchMemorySaver* current_saver_ = nullptr;
+
+    void set_current_saver(TorchMemorySaver* saver) {
+        current_saver_ = saver;
+    }
+
+    TorchMemorySaver* get_current_saver() {
+        return current_saver_;
+    }
 }
 
 // ------------------------------------------------- entrypoints ------------------------------------------------
 
 cudaError_t cudaMalloc(void **ptr, size_t size) {
-    if (RegionManager::is_interesting_region()) {
-        return TorchMemorySaver::instance().malloc(ptr, size);
+    TorchMemorySaver* saver = RegionManager::get_current_saver();
+    if (saver != nullptr && saver->is_interesting_region()) {
+        return saver->malloc(ptr, size);
     } else {
         return APIForwarder::call_real_cuda_malloc(ptr, size);
     }
 }
 
 cudaError_t cudaFree(void *ptr) {
-    if (RegionManager::is_interesting_region()) {
-        return TorchMemorySaver::instance().free(ptr);
+    TorchMemorySaver* saver = RegionManager::get_current_saver();
+    if (saver != nullptr && saver->is_interesting_region()) {
+        return saver->free(ptr);
     } else {
         return APIForwarder::call_real_cuda_free(ptr);
     }
 }
 
 extern "C" {
-void tms_region_enter() {
-    RegionManager::enter();
+void tms_region_enter(TorchMemorySaver* saver) {
+    if (saver == nullptr) {
```

nit: fail fast and just exit(1) for this, since this should never happen

(same for all others)

---

done